From ac17951bfd202f66fb849c5346a55539dc93f220 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:04:45 -0400
Subject: [ethernet] qlcnic: Back off patch for the upstream commit "d6e9c89a8d3cf0a5184badbcd50169179af27721"

Message-id: <1408626422-3941-2-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88541
O-Subject: [PATCH RHEL 7.1 001/138] qlcnic: Back off patch for the upstream commit "d6e9c89a8d3cf0a5184badbcd50169179af27721"
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Manish Chopra <manish.chopra@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

    qlcnic: Fix ethtool statistics length calculation

    o Consider number of Tx queues while calculating the length of
      Tx statistics as part of ethtool stats.
    o Calculate statistics lenght properly for 82xx and 83xx adapter

    Signed-off-by: Shahed Shaikh <shahed.shaikh@qlogic.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

o The above upstream commit was already ported by Redhat for BZ-1099634 which has some differences
  with upstream patch because few patches were not ported previous to this patch.
  In order to have this patch similar to upstream we need to backport
  some more patches from upstream and then after we need to backport the patch belonging to
  this commit "d6e9c89a8d3cf0a5184badbcd50169179af27721".
  This would make backporting easy and symmetrical so backing off the changes and will port
  the same upstream patch in [PATCH 062/138].

Signed-off-by: Manish Chopra <manish.chopra@qlogic.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
index 468c0ae..84023b8 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
@@ -164,35 +164,27 @@ static const char qlcnic_gstrings_test[][ETH_GSTRING_LEN] = {
 
 #define QLCNIC_TEST_LEN ARRAY_SIZE(qlcnic_gstrings_test)
 
-static inline int qlcnic_82xx_statistics(struct qlcnic_adapter *adapter)
+static inline int qlcnic_82xx_statistics(void)
 {
- return ARRAY_SIZE(qlcnic_gstrings_stats) +
-        ARRAY_SIZE(qlcnic_83xx_mac_stats_strings) +
-        ARRAY_SIZE(qlcnic_tx_ring_stats_strings) * adapter->max_drv_tx_rings;
+ return ARRAY_SIZE(qlcnic_device_gstrings_stats) +
+        ARRAY_SIZE(qlcnic_83xx_mac_stats_strings);
 }
 
-static inline int qlcnic_83xx_statistics(struct qlcnic_adapter *adapter)
+static inline int qlcnic_83xx_statistics(void)
 {
- return ARRAY_SIZE(qlcnic_gstrings_stats) +
-        ARRAY_SIZE(qlcnic_83xx_tx_stats_strings) +
+ return ARRAY_SIZE(qlcnic_83xx_tx_stats_strings) +
         ARRAY_SIZE(qlcnic_83xx_mac_stats_strings) +
-        ARRAY_SIZE(qlcnic_83xx_rx_stats_strings) +
-        ARRAY_SIZE(qlcnic_tx_ring_stats_strings) * adapter->max_drv_tx_rings;
+        ARRAY_SIZE(qlcnic_83xx_rx_stats_strings);
 }
 
 static int qlcnic_dev_statistics_len(struct qlcnic_adapter *adapter)
 {
- int len = -1;
-
- if (qlcnic_82xx_check(adapter)) {
-  len = qlcnic_82xx_statistics(adapter);
-  if (adapter->flags & QLCNIC_ESWITCH_ENABLED)
-   len += ARRAY_SIZE(qlcnic_device_gstrings_stats);
- } else if (qlcnic_83xx_check(adapter)) {
-  len = qlcnic_83xx_statistics(adapter);
- }
-
- return len;
+ if (qlcnic_82xx_check(adapter))
+  return qlcnic_82xx_statistics();
+ else if (qlcnic_83xx_check(adapter))
+  return qlcnic_83xx_statistics();
+ else
+  return -1;
 }
 
 #define QLCNIC_RING_REGS_COUNT 20
@@ -862,13 +854,18 @@ static int qlcnic_eeprom_test(struct net_device *dev)
 
 static int qlcnic_get_sset_count(struct net_device *dev, int sset)
 {
+ int len;
 
  struct qlcnic_adapter *adapter = netdev_priv(dev);
  switch (sset) {
  case ETH_SS_TEST:
   return QLCNIC_TEST_LEN;
  case ETH_SS_STATS:
-  return qlcnic_dev_statistics_len(adapter);
+  len = qlcnic_dev_statistics_len(adapter) + QLCNIC_STATS_LEN;
+  if ((adapter->flags & QLCNIC_ESWITCH_ENABLED) ||
+      qlcnic_83xx_check(adapter))
+   return len;
+  return qlcnic_82xx_statistics();
  default:
   return -EOPNOTSUPP;
  }
-- 
1.7.1